// react-geosuggest-plus/dist/GeosuggestItem.js
/* eslint react/jsx-sort-prop-types: 0, react/sort-comp: 0, react/prop-types: 0 */
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var GeosuggestItem = (function (_React$Component) {
  _inherits(GeosuggestItem, _React$Component);

  function GeosuggestItem() {
    _classCallCheck(this, GeosuggestItem);

    _get(Object.getPrototypeOf(GeosuggestItem.prototype), 'constructor', this).apply(this, arguments);
  }

  _createClass(GeosuggestItem, [{
    key: 'handleClick',

    /**
     * When the element gets clicked
     * @param  {Event} event The click event
     */
    value: function handleClick(event) {
      event.preventDefault();
      this.props.onSuggestSelect(this.props.suggest);
    }

    /**
     * The classes for the suggest item
     * @return {String} The classes
     */
  }, {
    key: 'getSuggestClasses',
    value: function getSuggestClasses() {
      var className = this.props.suggest.className;
      var classes = 'geosuggest-item';

      classes += this.props.isActive ? ' geosuggest-item--active' : '';
      classes += className ? ' ' + className : '';

      return classes;
    }
  }, {
    key: 'renderSaved',
    value: function renderSaved() {
      return _react2['default'].createElement(
        'li',
        { className: this.getSuggestClasses(), onClick: this.handleClick.bind(this) },
        _react2['default'].createElement('span', { className: 'icon icon-house' }),
        _react2['default'].createElement(
          'strong',
          null,
          this.props.suggest.firstname,
          ' ',
          this.props.suggest.lastname
        ),
        _react2['default'].createElement('br', null),
        this.props.suggest.place_name,
        this.props.suggest.address1,
        this.props.suggest.address2 ? ' ' + this.props.suggest.address2 : '',
        ',',
        ' ',
        this.props.suggest.city,
        ', ',
        this.props.suggest.state_name
      );
    }
  }, {
    key: 'renderGoogle',
    value: function renderGoogle() {
      return _react2['default'].createElement(
        'li',
        { className: this.getSuggestClasses(), onClick: this.handleClick.bind(this) },
        this.props.suggest.label
      );
    }

    /**
     * Render the view
     * @return {Function} The React element to render
     */
  }, {
    key: 'render',
    value: function render() {
      return this.props.suggest.label ? this.renderGoogle() : this.renderSaved();
    }
  }], [{
    key: 'propTypes',
    value: {
      isActive: _react2['default'].PropTypes.bool,
      suggest: _react2['default'].PropTypes.object,
      onSuggestSelect: _react2['default'].PropTypes.func
    },
    enumerable: true
  }]);

  return GeosuggestItem;
})(_react2['default'].Component);

GeosuggestItem.defaultProps = {
  isActive: false,
  suggest: {
    label: ''
  },
  onSuggestSelect: function onSuggestSelect() {}
};

exports['default'] = GeosuggestItem;
module.exports = exports['default'];
